das/dastree/dastree.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package dastree
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/offchainlabs/nitro/util/arbmath"
+)
+
+const BinSize = 64 * 1024 // 64 kB
+const NodeByte = byte(0xff)
+const LeafByte = byte(0xfe)
+
+type bytes32 = common.Hash
+
+type node struct {
+	hash bytes32
+	size uint32
+}
+
+func RecordHash(record func(bytes32, []byte), preimage ...[]byte) bytes32 {
+	// Algorithm
+	//  1. split the preimage into 64kB bins and double hash them to produce the tree's leaves
+	//  2. repeatedly hash pairs and their combined length, bubbling up any odd-one's out, to form the root
+	//
+	//            r         <=>  H(0xff, H(0xff, 0, 1, L(0:1)), 2, L(0:2))     step 3
+	//           / \
+	//          *   2       <=>  H(0xff, 0, 1, L(0:1)), 2                      step 2
+	//         / \
+	//        0   1         <=>  0, 1, 2                                       step 1
+	//
+	//      0   1   2       <=>  leaf n = H(0xfe, H(bin n))                    step 0
+	//
+	//  Where H is keccak and L is the length
+	//  Intermediate hashes like '*' from above may be recorded via the `record` closure
+	//
+
+	keccord := func(value []byte) bytes32 {
+		hash := crypto.Keccak256Hash(value)
+		record(hash, value)
+		return hash
+	}
+	prepend := func(before byte, slice []byte) []byte {
+		return append([]byte{before}, slice...)
+	}
+
+	unrolled := arbmath.ConcatByteSlices(preimage...)
+	if len(unrolled) == 0 {
+		return keccord(prepend(LeafByte, keccord([]byte{}).Bytes()))
+	}
+
+	length := uint32(len(unrolled))
+	leaves := []node{}
+	for bin := uint32(0); bin < length; bin += BinSize {
+		end := arbmath.MinUint32(bin+BinSize, length)
+		hash := keccord(prepend(LeafByte, keccord(unrolled[bin:end]).Bytes()))
+		leaves = append(leaves, node{hash, end - bin})
+	}
+
+	layer := leaves
+	for len(layer) > 1 {
+		prior := len(layer)
+		after := prior/2 + prior%2
+		paired := make([]node, after)
+		for i := 0; i < prior-1; i += 2 {
+			firstHash := layer[i].hash.Bytes()
+			otherHash := layer[i+1].hash.Bytes()
+			sizeUnder := layer[i].size + layer[i+1].size
+			dataUnder := firstHash
+			dataUnder = append(dataUnder, otherHash...)
+			dataUnder = append(dataUnder, arbmath.Uint32ToBytes(sizeUnder)...)
```

Comment:
Looks like a good opportunity for `ConcatByteSlices`

Reply:
Used!